Jenkinsfile

```diff
@@ -0,0 +1,542 @@
+ *
+ * ### Job configuration
+ *
+ * This Jenkinsfile gets its configuration from four sources:
+ * branch name, environment variables, a configuration file, and credentials.
+ * All configuration is optional for the default build (and it should stay that way),
+ * but some features require some configuration.
+ *
+ * #### Branch name
+ *
+ * See the org.hibernate.(...) imports for a link to the helpers library documentation,
+ * which explains the basics.
+ *
+ * #### Environment variables
+ *
+ * No particular environment variables is necessary.
+ *
+ * #### Job configuration file
+ *
+ * See the org.hibernate.(...) imports for a link to the helpers library documentation,
+ * which explains the basic structure of this file and how to set it up.
+ *
+ * Below is the additional structure specific to this Jenkinsfile:
+ *
+ *     deployment:
+ *       maven:
+ *         # String containing the ID of a Maven settings file registered using the config-file-provider Jenkins plugin.
+ *         # The settings must provide credentials to the servers with ID
+ *         # 'jboss-releases-repository' and 'jboss-snapshots-repository'.
+ *         settingsId: ...
+ */
+
+@Field final String MAVEN_TOOL = 'Apache Maven 3.6'
+
+// Default node pattern, to be used for resource-intensive stages.
+// Should not include the master node.
+@Field final String NODE_PATTERN_BASE = 'Slave'
+// Quick-use node pattern, to be used for very light, quick, and environment-independent stages,
+// such as sending a notification. May include the master node in particular.
+@Field final String QUICK_USE_NODE_PATTERN = 'Master||Slave'
+
+@Field AlternativeMultiMap<BuildEnvironment> environments
+@Field JobHelper helper
+
+@Field boolean enableDefaultBuild = false
+@Field boolean enableDefaultBuildIT = false
+@Field boolean performRelease = false
+@Field boolean deploySnapshot = false
+
+@Field Version releaseVersion
+@Field Version afterReleaseDevelopmentVersion
+
+this.helper = new JobHelper(this)
+
+helper.runWithNotification {
+
+stage('Configure') {
+	this.environments = AlternativeMultiMap.create([
+			jdk: [
+					// This should not include every JDK; in particular let's not care too much about EOL'd JDKs like version 9
+					// See http://www.oracle.com/technetwork/java/javase/eol-135779.html
+					new JdkBuildEnvironment(version: '8', buildJdkTool: 'OracleJDK8 Latest',
+							condition: TestCondition.BEFORE_MERGE,
+							isDefault: true),
+					new JdkBuildEnvironment(version: '11', buildJdkTool: 'OpenJDK 11 Latest',
+							condition: TestCondition.AFTER_MERGE),
+					new JdkBuildEnvironment(version: '14', buildJdkTool: 'OpenJDK 14 Latest',
+							condition: TestCondition.AFTER_MERGE),
+					new JdkBuildEnvironment(version: '15', buildJdkTool: 'OpenJDK 15 Latest',
+							condition: TestCondition.AFTER_MERGE),
+					new JdkBuildEnvironment(version: '16', buildJdkTool: 'OpenJDK 16 Latest',
+							condition: TestCondition.AFTER_MERGE)
```

Comment:
I was wondering if we should include the sigtest, jqassistant and the WildFly TCK container build (for JDK 8 and 14).

I'm not exactly sure they make sense for the main build as at least the TCK builds are quite slow but if they can be parallelized, maybe it makes sense.

This is an honest question, happy to have your feedback/benefit from your experience on this.